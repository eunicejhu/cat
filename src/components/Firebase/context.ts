import { createContext } from "react";
const FirebaseContext = createContext<unknown>(null);

export default FirebaseContext;
