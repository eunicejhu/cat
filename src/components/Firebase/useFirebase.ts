import { useContext } from "react";
import FirebaseContext from "./context";

export default function useFirebase() {
    const firebase = useContext(FirebaseContext);
    return firebase;
}
