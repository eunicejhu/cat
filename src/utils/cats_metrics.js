const R = require("ramda");

const average = (arr) =>
  Math.round((arr.reduce((sum, curr) => sum + curr, 0) / arr.length) * 100) /
  100;
const format = (str) => str.split("-").map((substr) => Number(substr.trim()));
const metrics = (data) =>
  data.map(({ weight: { metric } }) => average(format(metric)));
const lifeSpans = (data) =>
  data.map(({ life_span: lifeSpan }) => average(format(lifeSpan)));

const avgWeight = R.compose(average, metrics);
const avgLifeSpan = R.compose(average, lifeSpans);
const totalBreeds = (arr) => arr.length;

export { avgWeight, avgLifeSpan, totalBreeds };
