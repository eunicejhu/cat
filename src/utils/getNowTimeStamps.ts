const getNowTimeStamp = () => new Date().toISOString();
export default getNowTimeStamp;
