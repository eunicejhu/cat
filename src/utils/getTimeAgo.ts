import { parseISO, formatDistanceToNow } from "date-fns";
const getTimeAgo = (dateStr: string | undefined): string => {
  if (typeof dateStr === "string") {
    return formatDistanceToNow(parseISO(dateStr));
  }
  return "";
};

export default getTimeAgo;
