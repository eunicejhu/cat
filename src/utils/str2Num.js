const str2Num = (str) => Number(str.split(",").join(""));
export default { str2Num };
