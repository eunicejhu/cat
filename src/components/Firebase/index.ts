import app from "firebase/app";
import "firebase/auth";
import FirebaseContext from "./context";
import useFirebase from "./useFirebase";

//TODO: .env.production, .env.development
const config = {
    apiKey: process.env.REACT_APP_API_KEY,
    authDomain: process.env.REACT_APP_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_DATABASE_URL,
    projectId: process.env.REACT_APP_PROJECT_ID,
    storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_APP_ID,
    measurementId: process.env.REACT_APP_MEASUREMENT_ID,
};

class Firebase {
    auth: app.auth.Auth;
    constructor() {
        app.initializeApp(config);
        this.auth = app.auth();
        this.auth.useEmulator("http://localhost:9099/");
    }

    /**
     * Auth API
     */
    doCreateUserWithEmailAndPassword = (email: string, password: string) =>
        this.auth.createUserWithEmailAndPassword(email, password);
    doSignInWithEmailAndPassword = (email: string, password: string) =>
        this.auth.signInWithEmailAndPassword(email, password);
    doSignOut = () => this.auth.signOut();
    doPasswordReset = (email: string) =>
        this.auth.sendPasswordResetEmail(email);
    doPasswordUpdate = (newPassword: string) =>
        this.auth.currentUser?.updatePassword(newPassword);
}

export default Firebase;
export { FirebaseContext, useFirebase };
