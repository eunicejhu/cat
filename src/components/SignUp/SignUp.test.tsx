import React from "react";
import { mocked } from "ts-jest/utils";
import { BrowserRouter } from "react-router-dom";
import SignUpPage, { SignUpForm, SignUpLink } from "./index";
import Firebase, { FirebaseContext, useFirebase } from "../Firebase/index";
import * as ROUTES from "../../constants/routes";
import user from "@testing-library/user-event";
import { render } from "@testing-library/react";

jest.mock("../Firebase/index");

describe("SignUpForm test", () => {
    beforeEach(() => {
        mocked(Firebase).mockClear();
    });
    it("signup successfully and redirect to home page after filling in username, email, password", async () => {
        mocked(useFirebase).mockReturnValue({
            doCreateUserWithEmailAndPassword: jest.fn(),
        });
        const ui = (
            <FirebaseContext.Provider value={new Firebase()}>
                <BrowserRouter>
                    <SignUpForm />
                </BrowserRouter>
            </FirebaseContext.Provider>
        );
        const { getByTestId, findByTestId } = render(ui);

        user.type(getByTestId("username"), "eunice");
        user.type(getByTestId("email"), "eunice@gamil.com");
        user.type(getByTestId("passwordOne"), "password**");
        user.type(getByTestId("passwordTwo"), "password**");
        user.click(getByTestId("submit"));

        expect(await findByTestId("email")).toHaveValue("");
        expect(window.location.pathname).toBe(ROUTES.HOME);
    });

    it("cannot submit when any of field is blank or passwordOne is not identical to passwordTwo", () => {
        const ui = (
            <FirebaseContext.Provider value={new Firebase()}>
                <BrowserRouter>
                    <SignUpForm />
                </BrowserRouter>
            </FirebaseContext.Provider>
        );
        const { getByTestId } = render(ui);
        const submitBtn = getByTestId("submit") as HTMLButtonElement;

        user.type(getByTestId("username"), "eunice");
        expect(submitBtn.disabled).toBeTruthy();

        user.type(getByTestId("email"), "eunice@gamil.com");
        user.type(getByTestId("passwordOne"), "password**");
        user.type(getByTestId("passwordTwo"), "password**LLL");
        user.click(submitBtn);

        expect(submitBtn.disabled).toBeTruthy();
        user.type(getByTestId("passwordTwo"), "password**");
        expect(submitBtn.disabled).toBeFalsy();
    });
    it("show error message when Firebase signup authentication failed", async () => {
        const mockDoCreateUserWithEmailAndPassword = jest
            .fn()
            .mockRejectedValueOnce({ message: "failed to create user" });
        mocked(useFirebase).mockImplementation(() => ({
            doCreateUserWithEmailAndPassword: mockDoCreateUserWithEmailAndPassword,
        }));

        const ui = (
            <FirebaseContext.Provider value={new Firebase()}>
                <BrowserRouter>
                    <SignUpForm />
                </BrowserRouter>
            </FirebaseContext.Provider>
        );
        const { getByTestId, findByText } = render(ui);

        user.type(getByTestId("username"), "eunice");
        user.type(getByTestId("email"), "eunice@gamil.com");
        user.type(getByTestId("passwordOne"), "password**");
        user.type(getByTestId("passwordTwo"), "password**");
        user.click(getByTestId("submit"));

        expect(Firebase).toHaveBeenCalledTimes(1);
        expect(mockDoCreateUserWithEmailAndPassword).toHaveBeenLastCalledWith(
            "eunice@gamil.com",
            "password**"
        );
        expect(await findByText(/failed to create user/i)).toBeInTheDocument();
    });
});

describe("SignUpLink test", () => {
    it("click to change url to /signup", () => {
        const ui = (
            <BrowserRouter>
                <SignUpLink />
            </BrowserRouter>
        );
        const { getByText } = render(ui);
        user.click(getByText(/Sign Up/i));
        expect(window.location.pathname).toBe(ROUTES.SIGN_UP);
    });
});

describe("SignUpPage test", () => {
    it("render correctly", () => {
        const { asFragment } = render(<SignUpPage />);
        expect(asFragment()).toMatchSnapshot();
    });
});
