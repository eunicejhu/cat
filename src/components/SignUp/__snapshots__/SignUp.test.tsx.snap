// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpPage test render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz iRNCZj"
  >
    <h1>
      SignUp
    </h1>
    <form
      class="sc-bdfBwQ iXdyiL"
    >
      <input
        data-testid="username"
        name="username"
        placeholder="username"
        type="text"
        value=""
      />
      <input
        data-testid="email"
        name="email"
        placeholder="email"
        type="email"
        value=""
      />
      <input
        data-testid="passwordOne"
        name="passwordOne"
        placeholder="password"
        type="password"
        value=""
      />
      <input
        data-testid="passwordTwo"
        name="passwordTwo"
        placeholder="comfirm password"
        type="password"
        value=""
      />
      <p />
      <button
        class="sc-dlfnbm YJdWz"
        data-testid="submit"
        disabled=""
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
